modules/sdk/src/main/scala/org/tessellation/sdk/domain/cluster/programs/Joining.scala

```diff
@@ -206,10 +217,31 @@ sealed abstract class Joining[F[_]: Async: GenUUID: SecurityProvider: KryoSerial
       _ <- nodeStorage.tryModifyState(NodeState.SessionStarted, stateAfterJoining).handleError(_ => ())
     } yield peer
 
+  private def validateWhitelisting(peer: PeerToJoin): F[Unit] =
+    whitelisting match {
+      case None => Applicative[F].unit
+      case Some(entries) =>
+        if (entries.contains(peer.id)) Applicative[F].unit else PeerNotWhitelisted(peer.id).raiseError[F, Unit]
+    }
+
   private def validateHandshake(registrationRequest: RegistrationRequest, remoteAddress: Option[Host]): F[Unit] =
     for {
       ip <- registrationRequest.ip.pure[F]
 
+      ownClusterId = clusterStorage.getClusterId
+
+      _ <- if (registrationRequest.clusterId == ownClusterId)
+        Applicative[F].unit
+      else ClusterIdDoesNotMatch.raiseError[F, Unit]
+
+      ownClusterSession <- clusterStorage.getClusterSession
+
+      _ <- ownClusterSession match {
+        case Some(session) if session === registrationRequest.clusterSession => Applicative[F].unit
+        case None                                                            => Applicative[F].unit
```

Comment:
Shouldn't we only allow empty session during an initial join?

Reply:
Maybe we should reflect that here somehow?

Reply:
I think that for the rejoin (or healthcheck auto rejoin) we would require the same cluster session